pkg/pki/ed25519/ed25519.go

```diff
@@ -0,0 +1,189 @@
+	var inputBuffer bytes.Buffer
+
+	if _, err := io.Copy(&inputBuffer, r); err != nil {
+		return nil, fmt.Errorf("unable to read ed25519 signature: %w", err)
+	}
+
+	inputString := inputBuffer.String()
+	signature, err := minisign.DecodeSignature(inputString)
+	if err != nil {
+		//try to parse as signify
+		lines := strings.Split(strings.TrimRight(inputString, "\n"), "\n")
+		if len(lines) != 2 {
+			return nil, fmt.Errorf("invalid signature provided: %v lines detected", len(lines))
+		}
+		sigBytes, b64Err := base64.StdEncoding.DecodeString(lines[1])
+		if b64Err != nil {
+			return nil, fmt.Errorf("invalid signature provided: base64 decoding failed")
+		}
+		if len(sigBytes) != ed25519.SignatureSize+10 {
+			return nil, fmt.Errorf("invalid signature provided: incorrect size %v detected", len(sigBytes))
+		}
+		copy(signature.Signature[:], sigBytes[10:])
+	}
+	s.signature = &signature
+	return &s, nil
+}
+
+// FetchSignature implements pki.Signature interface
+func FetchSignature(ctx context.Context, url string) (*Signature, error) {
+	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error initializing fetch for ed25519 signature: %w", err)
+	}
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("error fetching ed25519 signature: %w", err)
+	}
+	defer resp.Body.Close()
+
+	sig, err := NewSignature(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	return sig, nil
+}
+
+// CanonicalValue implements the pki.Signature interface
+func (s Signature) CanonicalValue() ([]byte, error) {
+	if s.signature == nil {
+		return nil, fmt.Errorf("ed25519 signature has not been initialized")
+	}
+
+	buf := bytes.NewBuffer([]byte("untrusted comment:\n"))
+	b64Buf := bytes.NewBuffer(s.signature.SignatureAlgorithm[:])
+	if _, err := b64Buf.Write(s.signature.KeyId[:]); err != nil {
+		return nil, fmt.Errorf("error canonicalizing ed25519 signature: %w", err)
+	}
+	if _, err := b64Buf.Write(s.signature.Signature[:]); err != nil {
+		return nil, fmt.Errorf("error canonicalizing ed25519 signature: %w", err)
+	}
+	if _, err := buf.WriteString(base64.StdEncoding.EncodeToString(b64Buf.Bytes())); err != nil {
+		return nil, fmt.Errorf("error canonicalizing ed25519 signature: %w", err)
+	}
+	return buf.Bytes(), nil
+}
+
+// Verify implements the pki.Signature interface
+func (s Signature) Verify(r io.Reader, k interface{}) error {
+	if s.signature == nil {
+		return fmt.Errorf("ed25519 signature has not been initialized")
+	}
+
+	key, ok := k.(*PublicKey)
+	if !ok {
+		return fmt.Errorf("cannot use Verify with a non-ed25519 key")
+	}
+	if key.key == nil {
+		return fmt.Errorf("ed25519 public key has not been initialized")
+	}
+
+	msg, err := ioutil.ReadAll(r)
+	if err != nil {
+		return fmt.Errorf("error reading message to verify signature: %w", err)
+	}
+
+	if ok := ed25519.Verify(ed25519.PublicKey(key.key.PublicKey[:]), msg, s.signature.Signature[:]); !ok {
```

Comment:
The difference is that the link you mention will only validate Minisign signatures. Since signify signatures are a subset of Minisign, the check here validates only the basic signature (not the global signature or a trusted comment).

Reply:
Ah, makes sense.